flax/nnx/nnx/bridge/wrappers.py

```diff
@@ -139,11 +167,90 @@ def nn_var_to_nnx_state(kp, v):
+
+  Example::
+
+    >>> from flax import linen as nn, nnx
+    >>> import jax
+    >>> model = nnx.bridge.NNXToLinen(nnx.Linear, args=(32, 64))
+    >>> x = jax.numpy.ones((1, 32))
+    >>> y, variables = model.init_with_output(jax.random.key(0), x)
+    >>> y.shape
+    (1, 64)
+    >>> variables['params']['kernel'].value.shape
+    (32, 64)
+    >>> # The static GraphDef of the underlying NNX module
+    >>> variables.keys()
+    dict_keys(['nnx', 'params'])
+    >>> type(variables['nnx']['graphdef'])
+    <class 'flax.nnx.nnx.graph.GraphDef'>
+
+  Args:
+    nnx_class: The NNX Module class (not instance!).
+    args: The arguments that normally would be passed in to create the NNX module.
+    kwargs: The keyword arguments that normally would be passed in to create the NNX module.
+    skip_rng: True if this NNX module doesn't need `rngs` arg during initialization (not common).
+
+  Returns:
+    A stateful NNX module that behaves the same as the wrapped Linen module.
+  """
+  nnx_class: tp.Callable[..., Module]
+  args: tp.Sequence = dataclasses.field(default_factory=list)
+  kwargs: tp.Mapping = dataclasses.field(default_factory=dict)
+  skip_rng: bool = False
+
+  def update_variables(self, module):
+    """Store the NNX module's graph def and state inside Linen module variables."""
+    gdef, state = nnx.split(module)
+    # Save the graph def.
+    if self.is_mutable_collection('nnx'):
+      self.put_variable('nnx', 'graphdef', gdef)
+    # Sort all the variable types.
+    types = set(jax.tree.leaves(
+      jax.tree.map(lambda x: x.type, state,
+                    is_leaf=lambda x: isinstance(x, nnx.VariableState))))
+    types = variableslib.sort_variable_types(types)
+    _, *state_by_types = nnx.split(module, *types)
+    # Each variable type goes to its own linen collection, and
+    # each attribute goes to its own linen variable
+    for typ, state in zip(types, state_by_types):
+      collection = variableslib.variable_type_name(typ)
+      if self.is_mutable_collection(collection):
+        for k, v in state.raw_mapping.items():
+          self.put_variable(collection, k, v)
+
+  @linen.compact
   def __call__(self, *args, **kwargs):
-    ...
+    # init codepath
+    if self.is_initializing():
+      module_kwargs = dict(self.kwargs)
+      if not self.skip_rng:
+        module_kwargs |= dict(rngs=nnx.Rngs(**linen_rngs_dict(self)))
+      module = self.nnx_class(*self.args, **module_kwargs)
+      self.update_variables(module)
+      return module(*args, **kwargs)
+
+    # apply codepath
+    gdef = self.get_variable('nnx', 'graphdef')
+    states = [State(state) for col, state in self.variables.items() if col != 'nnx']
+    nnx_state = nnx.GraphState.merge(*states)
+    module = nnx.merge(gdef, nnx_state)
+    nnx.reseed(module, **linen_rngs_dict(self))  # reseed with keys from linen apply call.
+    out = module(*args, **kwargs)
+    self.update_variables(module)
+    return out
```

Comment:
We need to run call `update_variables` after calling the Module during initialization. We could refactor the code such that each branch (init / apply) constructs `module`:
```suggestion
    if self.is_initializing():
      module_kwargs = dict(self.kwargs)
      if not self.skip_rng:
        module_kwargs |= dict(rngs=nnx.Rngs(**linen_rngs_dict(self)))
      module = self.nnx_class(*self.args, **module_kwargs)
    else: # apply codepath
      gdef = self.get_variable('nnx', 'graphdef')
      states = [State(state) for col, state in self.variables.items() if col != 'nnx']
      nnx_state = nnx.GraphState.merge(*states) states else nnx.GraphState({})
      module = nnx.merge(gdef, nnx_state)
    nnx.reseed(module, **linen_rngs_dict(self))  # reseed with keys from linen apply call.
    out = module(*args, **kwargs)
    self.update_variables(module)
    return out
```
Added a small fix of checking there is at least 1 State for `merge`.

Reply:
Thanks for the `merge` fix - applied.
However, updating all variables in init time create divergent behavior from the NNX modules. For example, classes like `Counter` (see the unit test) will get +1 on Linen `init()` time, while remain 0 on NNX initialization.
I think that's not desired behavior. If people want the NNX state change after an NNX `__call__` run, they should run Linen `apply()` once and grab the updates.

Reply:
You are right! I think the previous behavior makes more sense.